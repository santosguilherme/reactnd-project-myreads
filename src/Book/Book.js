import React from 'react';
import PropTypes from 'prop-types';

import BookshelfChanger from '../BookshelfChanger/BookshelfChanger';

import './Book.css';


const Book = ({book, onUpdateBook}) => {
    const handleBookshelfChanger = shelf => {
        onUpdateBook({...book, shelf});
    };
    const {
        title,
        shelf,
        authors = [],
        imageLinks
    } = book;
    const bookCoverStyle = {width: 128, height: 193, backgroundColor: '#ccc'};

    if (imageLinks) {
        delete bookCoverStyle.backgroundColor;
        bookCoverStyle.backgroundImage = `url("${imageLinks.thumbnail}")`;
    }

    return (
        <div className="book">
            <div className="book-top">
                <div className="book-cover" style={bookCoverStyle}/>
                <BookshelfChanger
                    shelf={shelf}
                    onChangeOption={handleBookshelfChanger}
                />
            </div>
            <div className="book-title">
                {title}
            </div>
            <div className="book-authors">
                {authors.map(author => (
                    <span key={author}>{author}</span>
                ))}
            </div>
        </div>
    );
};

Book.propTypes = {
    book: PropTypes.object.isRequired,
    onUpdateBook: PropTypes.func.isRequired
};

export default Book;