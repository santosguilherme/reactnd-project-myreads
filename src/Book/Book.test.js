import React from 'react';
import {shallow, mount} from 'enzyme';

import Book from './Book';

it('renders without crashing', () => {
    const bookMock = {
        title: 'My Book',
        shelf: 'read',
        authors: ['Santos, Guilherme']
    };
    const onUpdateBookMock = jest.fn();

    const wrapper = shallow(
        <Book
            book={bookMock}
            onUpdateBook={onUpdateBookMock}
        />
    );

    expect(wrapper).toMatchSnapshot();
});

it('Should call the onUpdateBook property when change bookshelf', () => {
    const bookMock = {
        title: 'My Book',
        shelf: 'read',
        authors: ['Santos, Guilherme']
    };
    const onUpdateBookMock = jest.fn();

    const test = mount(
        <Book
            book={bookMock}
            onUpdateBook={onUpdateBookMock}
        />
    );

    test.find('select').simulate('change');
    expect(onUpdateBookMock).toHaveBeenCalledTimes(1);
});

it('Should render the backgroundColor style property when book hasnt a imageLink attribute', () => {
    const bookMock = {
        title: 'My Book',
        shelf: 'read',
        authors: ['Santos, Guilherme']
    };
    const onUpdateBookMock = jest.fn();

    const test = mount(
        <Book
            book={bookMock}
            onUpdateBook={onUpdateBookMock}
        />
    );

    const bookCoverStyle = test.find('.book-cover').prop('style');
    const {backgroundImage, backgroundColor} = bookCoverStyle;

    expect(backgroundColor).toBeDefined();
    expect(backgroundImage).toBeUndefined();
});

it('Should render the backgroundImage style property when book has a imageLink attribute', () => {
    const thumbnail = 'http://lorempixel.com/128/193';
    const bookMock = {
        title: 'My Book',
        shelf: 'read',
        authors: ['Santos, Guilherme'],
        imageLinks: {thumbnail}
    };
    const onUpdateBookMock = jest.fn();

    const test = mount(
        <Book
            book={bookMock}
            onUpdateBook={onUpdateBookMock}
        />
    );

    const bookCoverStyle = test.find('.book-cover').prop('style');
    const {backgroundImage, backgroundColor} = bookCoverStyle;

    expect(backgroundImage).toMatch(new RegExp(thumbnail));
    expect(backgroundColor).toBeUndefined();
});
