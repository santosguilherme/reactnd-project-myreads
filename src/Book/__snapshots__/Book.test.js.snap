// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Book
    book={
        Object {
            "authors": Array [
              "Santos, Guilherme",
            ],
            "shelf": "read",
            "title": "My Book",
          }
    }
    onUpdateBook={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="book-top"
>
          <div
                    className="book-cover"
                    style={
                              Object {
                                        "backgroundColor": "#ccc",
                                        "height": 193,
                                        "width": 128,
                                      }
                    }
          />
          <BookshelfChanger
                    onChangeOption={[Function]}
                    shelf="read"
          />
</div>,
        <div
          className="book-title"
>
          My Book
</div>,
        <div
          className="book-authors"
>
          <span>
                    Santos, Guilherme
          </span>
</div>,
      ],
      "className": "book",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="book-cover"
              style={
                            Object {
                                          "backgroundColor": "#ccc",
                                          "height": 193,
                                          "width": 128,
                                        }
              }
/>,
            <BookshelfChanger
              onChangeOption={[Function]}
              shelf="read"
/>,
          ],
          "className": "book-top",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "className": "book-cover",
              "style": Object {
                "backgroundColor": "#ccc",
                "height": 193,
                "width": 128,
              },
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "onChangeOption": [Function],
              "shelf": "read",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "My Book",
          "className": "book-title",
        },
        "ref": null,
        "rendered": "My Book",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span>
              Santos, Guilherme
</span>,
          ],
          "className": "book-authors",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "Santos, Guilherme",
            "nodeType": "host",
            "props": Object {
              "children": "Santos, Guilherme",
            },
            "ref": null,
            "rendered": "Santos, Guilherme",
            "type": "span",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="book-top"
>
            <div
                        className="book-cover"
                        style={
                                    Object {
                                                "backgroundColor": "#ccc",
                                                "height": 193,
                                                "width": 128,
                                              }
                        }
            />
            <BookshelfChanger
                        onChangeOption={[Function]}
                        shelf="read"
            />
</div>,
          <div
            className="book-title"
>
            My Book
</div>,
          <div
            className="book-authors"
>
            <span>
                        Santos, Guilherme
            </span>
</div>,
        ],
        "className": "book",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="book-cover"
                style={
                                Object {
                                                "backgroundColor": "#ccc",
                                                "height": 193,
                                                "width": 128,
                                              }
                }
/>,
              <BookshelfChanger
                onChangeOption={[Function]}
                shelf="read"
/>,
            ],
            "className": "book-top",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "className": "book-cover",
                "style": Object {
                  "backgroundColor": "#ccc",
                  "height": 193,
                  "width": 128,
                },
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "onChangeOption": [Function],
                "shelf": "read",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "My Book",
            "className": "book-title",
          },
          "ref": null,
          "rendered": "My Book",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span>
                Santos, Guilherme
</span>,
            ],
            "className": "book-authors",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Santos, Guilherme",
              "nodeType": "host",
              "props": Object {
                "children": "Santos, Guilherme",
              },
              "ref": null,
              "rendered": "Santos, Guilherme",
              "type": "span",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
